### ISSUE TITLE
  Inconsistency in Results Using sum(. != 3) within Pipeline Operations (%>%)  dplyr version 1.1.4

### ISSUE THREAD

**Issue by ZHBHSMILE** [tidyverse/dplyr#7037]
I encountered an inconsistency in the results obtained from different approaches while working with R code. Specifically, when using the sum(. != 3) expression within a pipeline operation (%>%), the outcome varied from expectations, leading to discrepancies compared to direct computations.  

```r
  data <- data.frame(
    GO.BiologicalProcess = c("-", "-", "A", "B"),
    GO.CellularComponent = c("-", "C", "-", "D"),
    GO.MolecularFunction = c("-", "-", "-", "E")
  )
  
  # Calculate the number of occurrences of "-" in each row
  go_num <- rowSums(data[, c("GO.BiologicalProcess", "GO.CellularComponent", "GO.MolecularFunction")] == "-")
  
  # Calculate the sum of occurrences where the count is not equal to 3
  go_bg_num <- sum(go_num != 3)
  
  # Using pipe operator and sum(. != 3)
  error_sum <- rowSums(data[, c("GO.BiologicalProcess", "GO.CellularComponent", "GO.MolecularFunction")] == "-") %>% sum(. != 3)
```

**Comment by SpatLyu** [comment]
Calculate it like this:
``` r
library(magrittr)

data <- data.frame(
  GO.BiologicalProcess = c("-", "-", "A", "B"),
  GO.CellularComponent = c("-", "C", "-", "D"),
  GO.MolecularFunction = c("-", "-", "-", "E")
)

{data[, c("GO.BiologicalProcess", "GO.CellularComponent", "GO.MolecularFunction")] == "-"} %>% 
  rowSums() %>% 
  {. != 3} %>% 
  sum()
#> [1] 3
```

<sup>Created on 2024-06-10 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by philibe** [comment]
TLDR: You're counting boolean. (In your second post).

initial datas
```R
  foo<-{data[, c("GO.BiologicalProcess", "GO.CellularComponent", "GO.MolecularFunction")] == "-"} %>% rowSums() 
  foo
  # [1] 3 2 2 0
```
`{. != 3}` make boolean :
```R
  foo2 <-  foo  %>%    {. != 3}
  foo2
  # [1] FALSE  TRUE  TRUE  TRUE
````

Therefore `sum` is number of `TRUE` = 3

At the opposite
```R
  foo %>%  {. != 3} %>% sum(.=TRUE)
  # [1] 4
````
And in basic R or basic R with `%>%`
```
  sum(foo[foo!=3] )
  # [1] 4
  foo[foo!=3] %>% sum(.)
  # [1] 4
```

or piped function or not piped function:
```R
  testsum <- function (x) if_else(x!=3,x,0)
  foo %>% testsum(.) %>% sum(.)
  # [1] 4
```

**Comment by philibe** [comment]
In your 1st post `sum(go_num != 3)` should be  `go_bg_num <- sum(go_num[go_num!=3] )` or `go_bg_num <- tibble( x = go_num) %>% filter(x!=3) %>% sum(.)`.

PS: I'm simple user.

**Comment by DavisVaughan** [comment]
Read ``` ?`%>%` ``` closely, particularly `Using the dot for secondary purposes` and you will see that this is known and expected behavior. The LHS is being placed as the first argument of `sum()`, _and_ as `.` in the second argument to `sum()` (the `. != 3` expression).

You can use braces to avoid this

```r
rowSums(data[, c("GO.BiologicalProcess", "GO.CellularComponent", "GO.MolecularFunction")] == "-") %>% {
  sum(. != 3)
}
```

but I think avoid the pipe entirely is cleaner here

```r
sums <- rowSums(data[, c("GO.BiologicalProcess", "GO.CellularComponent", "GO.MolecularFunction")] == "-")
sum <- sum(sums != 3)
```

**Comment by philibe** [comment]
Thanks for the explanation. I've looked for why  `c(3, 2, 2, 0) %>% sum(.!=3)`=10, which seemed very weird. :)
